src/ophyd_async/core/_signal.py

```diff
@@ -154,29 +154,28 @@ def _callback(self, reading: Reading[SignalDatatypeT]) -> None:
         )
         self._reading = reading
         self._valid.set()
-        items = self._listeners.copy().items()
-        for function, want_value in items:
-            self._notify(function, want_value)
+        callbacks = self._listeners.copy()
+        for callback in callbacks:
```

Comment:
```suggestion
        # Copy the listeners in case one of the callbacks removes the listener from the set
        for callback in list(self._listeners):
```